Added `kafka::rpc` method to get partition offsets

mmaslankaprv opened the pull request:
Added an API reassembling Kafka's `list_offsets` API but for internal Redpanda partitions. Using `kafka::data::rpc::client` allows callers to request offsets from multiple partitions that can have leaders on different nodes in cluster. The client aggregates the partition and dispatch requests to correct partition leaders.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
k/rpc: added request/response pair for list offsets kafka rpc

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
k/rpc: added kafka rpc method to get partition offsets

Added an API reassembling Kafka's `list_offsets` API but for internal
Redpanda partitions. Using `kafka::data::rpc::client` allows callers to
request offsets from multiple partitions that can have leaders on
different nodes in cluster. The client aggregates the partition and
dispatch requests to correct partition leaders.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
k/data/rpc: added basic test for listing partition offsets

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/dt

mmaslankaprv commented:
/dt

michael-redpanda commented on src/v/kafka/data/rpc/service.cc:
```diff
@@ -91,6 +91,67 @@ ss::future<ss::chunked_fifo<kafka_topic_data_result>> local_service::produce(
+        co_await ss::max_concurrent_for_each(
+          t.partitions.begin(),
+          t.partitions.end(),
+          32,
```
nit: should this be a constant defined somewhere?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--307f9825-bcc3-4309-8f76-d497d33c4cc2-->
<!--TEST RESULTS-->
<!--6db76980-98e1-44f1-a569-07ff179821c6-->
<!--TEST RESULTS-->
<!--7b4fc590-d152-4b59-83bf-d4f724de729d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70290">build#70290</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70290#01987eb3-8ab8-4824-9637-e14c94894bb5 | FLAKY         | 20/21    | upstream reliability is '99.53488372093024'. current run reliability is '95.23809523809523'. drift is 4.29679 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70299">build#70299</a></summary>


| test_class                       | test_method                          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------------|:-------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest            | test_creating_and_listing_migrations | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/70299#01987fb2-375c-4435-b9b4-82bff7779f8d | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             |
| DataMigrationsApiTest            | test_higher_level_migration_api      | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/70299#01987faf-b083-4169-838d-e5803c3b3b68 | FLAKY         | 15/21    | upstream reliability is '96.64694280078896'. current run reliability is '71.42857142857143'. drift is 25.21837 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest         | test_node_operations                 | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70299#01987faf-b085-4c2c-a661-b9fa3d532624 | FLAKY         | 19/21    | upstream reliability is '99.55947136563876'. current run reliability is '90.47619047619048'. drift is 9.08328 and the allowed drift is set to 50. The test should PASS  |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss           | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/70299#01987faf-b084-4bc9-b96f-c87d23c2f45c | FLAKY         | 20/21    | upstream reliability is '99.03614457831326'. current run reliability is '95.23809523809523'. drift is 3.79805 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70377">build#70377</a></summary>


| test_class     | test_method      | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:---------------|:-----------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| OIDCReauthTest | test_oidc_reauth | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70377#019883f0-53d9-451f-859d-177fac53c4d3 | FLAKY         | 20/21    | upstream reliability is '99.7134670487106'. current run reliability is '95.23809523809523'. drift is 4.47537 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

michael-redpanda reviewed (dismissed):
Just one nit, otherwise this looks great to me :) 🔥

oleiman commented on src/v/kafka/data/rpc/test/kafka_data_rpc_test.cc:
```diff
@@ -206,6 +220,33 @@ TEST_P(KafkaDataRpcTest, ClientCanUpdateTopics) {
+      model::ktp(not_existing.tp.topic, model::partition_id(0)));
+    ASSERT_TRUE(ne_res.has_value());
+    auto offsets_2 = std::move(ne_res.value());
+    EXPECT_EQ(offsets_2.size(), 1);
```
there should be an error in there right?

oleiman replied:
I meant 'should we check for the error', but i suppose it's fine

oleiman reviewed (commented):
lgtm modulo one q about the test case for topic_not_exists

mmaslankaprv committed:
k/rpc: wire in topic metadata cache in the client

Wired in the topic configuration cache in the client to make sure that
client is able to verify if the topic exists before querying the
leadership information.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

dotnwat commented:
just wondering what feature this is for?

oleiman commented:
> just wondering what feature this is for?

DR
